### PR TITLE
history panel

### PR DESCRIPTION
**Pull request by StephenHodgson**

### CODE REVIEW

File: Runtime/ApiRequests.cs [hunk 1]
```diff
@@ -1,24 +1,30 @@
...
+        private static async Task<T> GetAsync<T>(string path, string apiKey, Dictionary<string, string> queryParams = null)
         {
-            var queryString = "?api_key=" + UnityWebRequest.EscapeURL(apiKey);
-            if (queryParams.Length > 0)
+            if (!string.IsNullOrWhiteSpace(apiKey))
             {
-                queryString += "&" + string.Join("&", queryParams.Select(kv =>
-                    UnityWebRequest.EscapeURL(kv.Item1) + "=" + UnityWebRequest.EscapeURL(kv.Item2)));
+                queryParams ??= new();
+                queryParams.Add("api_key", UnityWebRequest.EscapeURL(apiKey));
```
Comment (afarchy) [inline Runtime/ApiRequests.cs:21]: you're double-escaping apiKey below

Comment (afarchy) [reply Runtime/ApiRequests.cs:21]: I didn't do it this way because I wanted apikey to be first

File: Runtime/ApiRequests.cs [hunk 2]
```diff
@@ -1,24 +1,30 @@
...
-        private static async Task<T> GetAsync<T>(string path, string apiKey, params (string, string)[] queryParams)
+        private static async Task<T> GetAsync<T>(string path, string apiKey, Dictionary<string, string> queryParams = null)
         {
-            var queryString = "?api_key=" + UnityWebRequest.EscapeURL(apiKey);
-            if (queryParams.Length > 0)
+            if (!string.IsNullOrWhiteSpace(apiKey))
             {
-                queryString += "&" + string.Join("&", queryParams.Select(kv =>
-                    UnityWebRequest.EscapeURL(kv.Item1) + "=" + UnityWebRequest.EscapeURL(kv.Item2)));
+                queryParams ??= new();
```
Comment (afarchy) [inline Runtime/ApiRequests.cs:20]: older version of Unity don't support new()

File: Runtime/ApiRequests.cs [hunk 3]
```diff
@@ -121,6 +128,73 @@ public static async Task<SkyboxTip> GetSkyboxTipAsync(string apiKey, SkyboxAiMod
             }
         }
 
+        public static async Task<GetHistoryResult> GetSkyboxHistoryAsync(string apiKey, HistorySearchQueryParameters searchQueryParams = null)
+        {
+            Dictionary<string, string> searchQuery = null;
+
+            if (searchQueryParams != null)
+            {
+                searchQuery = new();
```
Comment (afarchy) [inline Runtime/ApiRequests.cs:143]: new()

File: Runtime/BlockadeLabsSkyboxMesh.cs [hunk 4]
```diff
@@ -123,7 +123,7 @@ internal GetImagineResult GetMetadata()
 
             if (renderer.sharedMaterial.mainTexture.name == "default_skybox_texture")
             {
-                new GetImagineResult() { request = new GetImagineRequest() { id = 0 } };
+                _ = new GetImagineResult { request = new ImagineResult { id = 0 } };
```
Comment (afarchy) [inline Runtime/BlockadeLabsSkyboxMesh.cs:126]: .... I think this is supposed to return

File: Runtime/HistoryItemBehaviour.cs [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            _likeToggle.SetIsOnWithoutNotify(result != null && result.request.isMyFavorite);
+        }
+
+        private void OnRemoveButtonClicked()
+        {
+            Debug.Log($"Remove {_imagineResult.id}");
+            // TODO show confirmation dialog
+        }
+
+        private void OnDownloadButtonClicked()
```
Comment (afarchy) [inline Runtime/HistoryItemBehaviour.cs:72]: should probably remove the download button

File: Runtime/HistoryItemBehaviour.cs [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+            {
+                // listen if mouse button click occurs and turn off the options toggle
+                StartCoroutine(ListenForMouseClick());
+            }
+        }
+
+        private IEnumerator ListenForMouseClick()
+        {
+            // wait one frame to make sure we don't close the menu
+            yield return new WaitForEndOfFrame();
```
Comment (afarchy) [inline Runtime/HistoryItemBehaviour.cs:89]: allocate WaitForEndOfFrame once

File: Runtime/JsonFields.cs [hunk 7]
```diff
@@ -32,22 +32,27 @@ internal class CreateSkyboxResult
...
         public string depth_map_url;
         public string status;
         public string error_message;
         public string prompt;
         public string negative_text;
         public int seed;
         public int skybox_style_id;
+        public string skybox_style_name;
+        public DateTime completed_at;
+        public bool isMyFavorite;
```
Comment (afarchy) [inline Runtime/JsonFields.cs:55]: is this not snake case in the json?

**Review by afarchy — APPROVED** [review]

File: Runtime/ApiRequests.cs (hunk 2 above)
Comment (StephenHodgson) [reply Runtime/ApiRequests.cs]: what is the min supported version?

File: Runtime/BlockadeLabsSkyboxMesh.cs (hunk 4 above)
Comment (StephenHodgson) [reply Runtime/BlockadeLabsSkyboxMesh.cs:126]: it was being discarded before, now it is explicit is all.

File: Runtime/JsonFields.cs (hunk 7 above)
Comment (StephenHodgson) [reply Runtime/JsonFields.cs:55]: is edge pascalCase 😉

File: Runtime/HistoryItemBehaviour.cs (hunk 5 above)
Comment (StephenHodgson) [reply Runtime/HistoryItemBehaviour.cs:93]: pretty sure we need to download and load it into the main skybox scene.

Comment (afarchy) [reply Runtime/HistoryItemBehaviour.cs:93]: I think in the website the purpose of this button is to download the PNG or whatever. I'm not sure if we need that functionality here, instead just click the whole thumbnail to download/open it in the scene.

File: Meshes/Tetrahedron.asset
Comment (afarchy) [inline Meshes/Tetrahedron.asset:1]: replace with low poly version, and maybe rename PreviewSkyboxMesh or something

File: PreviewRenderTexture.renderTexture [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+  m_ObjectHideFlags: 0
+  m_CorrespondingSourceObject: {fileID: 0}
+  m_PrefabInstance: {fileID: 0}
+  m_PrefabAsset: {fileID: 0}
+  m_Name: PreviewRenderTexture
+  m_ImageContentsHash:
+    serializedVersion: 2
+    Hash: 00000000000000000000000000000000
+  m_ForcedFallbackFormat: 4
+  m_DownscaleFallback: 0
```
Comment (afarchy) [inline PreviewRenderTexture.renderTexture:14]: move to a folder

File: Runtime/ApiRequests.cs [hunk 9]
```diff
@@ -109,16 +125,129 @@ public static async Task DownloadFileAsync(string url, string path)
...
+                }
             }
+
+            return await GetAsync<GetHistoryResult>("imagine/myRequests", searchQuery);
+        }
+
+        public static async Task<ImagineResult> ToggleFavorite(int imagineId)
+        {
+#if UNITY_2022_1_OR_NEWER
+            using var request = UnityWebRequest.PostWwwForm($"{ApiEndpoint}imagine/favorite/{imagineId}", null);
```
Comment (afarchy) [inline Runtime/ApiRequests.cs:206]: I'm curious why we don't just always use the other version

File: Runtime/BlockadeDemoCamera.cs [hunk 10]
```diff
@@ -212,6 +249,11 @@ private void UpdateAutoPanning()
 
         private bool MouseIsOverUI()
         {
+            if (_rectTransform)
+            {
+                return !RectTransformUtility.RectangleContainsScreenPoint(_rectTransform, Input.mousePosition);
```
Comment (afarchy) [inline Runtime/BlockadeDemoCamera.cs:254]: this seems like a hack, or at least, can we rename the method?

File: Runtime/ExtensionMethods.cs [hunk 11]
```diff
@@ -0,0 +1,12 @@
+﻿using UnityEngine.Networking;
+
+namespace BlockadeLabsSDK
+{
+    internal static class AwaiterExtensions
```
Comment (afarchy) [inline Runtime/ExtensionMethods.cs:5]: why move this?

File: Runtime/HideAfterClick.cs [hunk 12]
```diff
@@ -0,0 +1,52 @@
+﻿using System.Collections;
+using UnityEngine;
+
+namespace BlockadeLabsSDK
+{
+    public class HideAfterClick : MonoBehaviour
```
Comment (afarchy) [inline Runtime/HideAfterClick.cs:6]: fancy

File: Runtime/HistoryItemBehaviour.cs [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+                _clickCallback = clickCallback;
+                _deleteCallback = deleteCallback;
+                _downloadCallback = downloadCallback;
+                _descriptionText.text = $"<b>{item.skybox_style_name}</b> | {item.prompt}";
+                _timestampText.text = item.completed_at.ToString("G");
+                _likeToggle.SetIsOnWithoutNotify(item.isMyFavorite);
+                _modelBadge.SetActive(item.model == "Model 3");
+                _apiBadge.SetActive(item.api_key_id != 0);
+
+                if (!_imageCache.TryGetValue(item.obfuscated_id, out var cachedImages))
```
Comment (afarchy) [inline Runtime/HistoryItemBehaviour.cs:187]: maybe ok for this scenario, but to be correct you need to add a blank item to the cache before downloading to ensure you don't download twice. See style picker preview

File: Runtime/HistoryPanel.cs [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+                    }
+                },
+                onCancel: () =>
+                {
+                    // empty callback will close dialog
+                });
+
+        private async void OnHistoryItemDownload(ImagineResult imagineResult)
+        {
+#if UNITY_EDITOR
```
Comment (afarchy) [inline Runtime/HistoryPanel.cs:157]: why only editor?

File: Runtime/HistoryItemBehaviour.cs (hunk 13 above)
Comment (afarchy) [reply Runtime/HistoryItemBehaviour.cs:187]: future: refactor image caching

**Review by afarchy — APPROVED** [review]
Looks good, just minor comments

File: Runtime/ApiRequests.cs (hunk 9 above)
Comment (StephenHodgson) [reply Runtime/ApiRequests.cs:206]: bc it is obsolete

File: Runtime/HistoryPanel.cs (hunk 14 above)
Comment (StephenHodgson) [reply Runtime/HistoryPanel.cs:157]: bc there may not be a place to download and use skybox at runtime (i.e. webgl)

only makes sense to view it.

File: Runtime/ExtensionMethods.cs (hunk 11 above)
Comment (StephenHodgson) [reply Runtime/ExtensionMethods.cs:5]: I prefer all classes/members and types to be in their own files per MSDN

File: Runtime/BlockadeDemoCamera.cs (hunk 10 above)
Comment (StephenHodgson) [reply Runtime/BlockadeDemoCamera.cs:254]: suggestions for a better name?

File: Runtime/HistoryItemBehaviour.cs (hunk 13 above)
Comment (StephenHodgson) [reply Runtime/HistoryItemBehaviour.cs:187]: I'm used to by rest package handling caching by default. Guess I've been spoiled 😆

**Pull request merged**